### PR TITLE
Add support for /session/reauth

### PR DESCRIPTION
**Pull request by rfk**
This adds a `.sessionReauth` method which is pretty straightforward, very similar to the existing `.signIn` method.

There's also a preparatory commit that rejiggers the way we do mock responses, to more effectively test the "incorrect email case" retry loop for both `/account/login` and `/session/reauth`.  These tests required the ability to have two mock responses in flight at once, because the client makes two requests during the function under test.  What I've done here *seems* to be OK but I don't really know this repo very well, so it might have unintended side-effects...

@mozilla/fxa-devs r?

### CODE REVIEW

File: tests/addons/environment.js [hunk 1]
```diff
@@ -30,13 +29,36 @@ define([
...
+      this.responses = [];
       // switch to the fake XHR
       this.xhr = Sinon.useFakeXMLHttpRequest();
       this.xhr.onCreate = function (xhr) {
+        if (self.requests.length < self.responses.length) {
+          var mock = self.responses[self.requests.length];
+          setTimeout(function() {
+            xhr.respond(mock.status, mock.headers, mock.body);
+          }, 0);
+        }
```
Comment (rfk) [inline tests/addons/environment.js]: This is new, and handles the case where the mock response was set up before the XHR request was created.

File: tests/addons/environment.js [hunk 2]
```diff
@@ -30,13 +29,36 @@ define([
...
+        if (typeof mock === 'undefined') {
+          console.log('Mock does not exist!');
+        }
+        // this has to be here to work in IE
+        setTimeout(function () {
+          if (self.responses.length < self.requests.length) {
+            self.requests[self.responses.length].respond(mock.status, mock.headers, mock.body);
+          }
+          self.responses.push(mock);
+        }, 0);
```
Comment (rfk) [inline tests/addons/environment.js]: This is similar to before, and handles the case where the XHR is already created when the mock response gets setup.

File: tests/addons/sinonResponder.js [hunk 3]
```diff
@@ -1,34 +0,0 @@
...
-    respond: function (req, mock) {
-      if (typeof mock === 'undefined') {
-        console.log('Mock does not exist!');
-      }
-      if (req) {
-        req.respond(mock.status, mock.headers, mock.body);
-      }
-    }
-  };
-});
```
Comment (rfk) [inline tests/addons/sinonResponder.js]: This code was inlined into its only caller.

File: tests/addons/sinonResponder.js [hunk 4]
```diff
@@ -1,34 +0,0 @@
...
-  'tests/addons/sinon'
-], function (Sinon) {
-
-  return {
-    makeMockResponder: function (requests) {
-      var self = this;
-      var requestIndex = 0;
-
-      return function (returnValue, response) {
-        Promise.resolve().then(function () {
```
Comment (rfk) [inline tests/addons/sinonResponder.js]: I couldn't figure out what this `Promise` for was, so I removed it...but something might break in CI to tell me what it was for? :-)

File: tests/lib/signIn.js [hunk 5]
```diff
@@ -158,35 +160,40 @@ define([
 
         return accountHelper.newVerifiedAccount()
           .then(function (account) {
+            var numSetupRequests = requests.length;
             var incorrectCaseEmail = account.input.email.charAt(0).toUpperCase() + account.input.email.slice(1);
 
-            return respond(client.signIn(incorrectCaseEmail, account.input.password), RequestMocks.signIn);
```
Comment (rfk) [inline tests/lib/signIn.js]: As far as I can tell, this test wasn't actually exercising the "incorrect case" code-path.

File: tests/mocks/errors.js [hunk 6]
```diff
@@ -122,7 +122,7 @@ define([], function () {
     incorrectEmailCase: {
       status: 400,
       headers: {},
-      body: '{"code":400, "errno": 120}'
+      body: '{"code":400, "errno": 120, "email": "a@b.com"}'
```
Comment (rfk) [inline tests/mocks/errors.js]: The "incorrect case" error only gets special treatment if it returns the correctly-cased email.

File: client/FxAccountClient.js [hunk 7]
```diff
@@ -905,6 +905,123 @@ define([
...
+   *   @param {Object} [options.metricsContext={}] Metrics context metadata
+   *     @param {String} options.metricsContext.deviceId identifier for the current device
+   *     @param {String} options.metricsContext.flowId identifier for the current event flow
+   *     @param {Number} options.metricsContext.flowBeginTime flow.begin event time
+   *     @param {Number} options.metricsContext.utmCampaign marketing campaign identifier
+   *     @param {Number} options.metricsContext.utmContent content identifier
+   *     @param {Number} options.metricsContext.utmMedium acquisition medium
+   *     @param {Number} options.metricsContext.utmSource traffic source
+   *     @param {Number} options.metricsContext.utmTerm search terms
+   *   @param {String} [options.unblockCode]
```
Comment (vbudhram) [inline client/FxAccountClient.js]: Looks like `.signIn` and this didn't include the `verificationMethod` and `originalLoginEmail` options.

File: tests/lib/session.js [hunk 8]
```diff
@@ -110,6 +112,113 @@ define([
...
+            var numSetupRequests = requests ? requests.length : null;
+            var sessionToken = account.signIn.sessionToken;
+            var incorrectCaseEmail = account.input.email.charAt(0).toUpperCase() + account.input.email.slice(1);
+            var password = account.input.password;
+
+            respond(ErrorMocks.incorrectEmailCase);
+            return respond(client.sessionReauth(sessionToken, incorrectCaseEmail, password), RequestMocks.sessionReauth)
+              .then(
+                function (res) {
+                  assert.property(res, 'uid');
```
Comment (vbudhram) [inline tests/lib/session.js]: nit: assert authAt and verified are set?

File: client/FxAccountClient.js [hunk 9]
```diff
@@ -905,6 +905,123 @@ define([
...
+            data.originalLoginEmail = options.originalLoginEmail;
+          }
+
+          if (options.verificationMethod) {
+            data.verificationMethod = options.verificationMethod;
+          }
+
+          return hawkCredentials(sessionToken, 'sessionToken',  HKDF_SIZE)
+            .then(function (creds) {
+              return self.request.send(endpoint, 'POST', creds, data);
```
Comment (vladikoff) [inline client/FxAccountClient.js]: nit: @rfk if you want EVEN more testing you can take a look at https://github.com/mozilla/fxa-js-client/pull/217/files to make sure the correct options are sent

Comment (rfk) [reply client/FxAccountClient.js]: Thanks!  I am *so* there for even more testing :-)

**Comment by rfk** [comment]
@vladikoff could you please give me new tests a look-over for completeness?

**Comment by rfk** [comment]
Yep, looks like the live server didn't like one of my made-up request option values, trying again...

File: tests/lib/session.js [hunk 10]
```diff
@@ -110,6 +122,216 @@ define([
...
+                function (res) {
+                  assert.ok(res.uid);
+                  assert.ok(res.verified);
+                  assert.ok(res.authAt);
+                  assert.ok(res.keyFetchToken);
+                  assert.ok(res.unwrapBKey);
+
+                  var args = xhr.prototype.open.args[xhr.prototype.open.args.length - 1];
+                  assert.equal(args[0], 'POST');
+                  assert.include(args[1], '/session/reauth?keys=true');
```
Comment (vladikoff) [inline tests/lib/session.js]: 👍

File: tests/lib/session.js [hunk 11]
```diff
@@ -110,6 +122,216 @@ define([
...
+
+                  var payload = JSON.parse(xhr.prototype.send.args[xhr.prototype.send.args.length - 1][0]);
```
Comment (vladikoff) [inline tests/lib/session.js]: 👍

File: tests/mocks/errors.js (hunk 6 above)
Comment (vladikoff) [reply tests/mocks/errors.js]: 👍

**Pull request merged**